### ISSUE TITLE
Default values and type incompatibilites

### ISSUE THREAD

**Issue by nilskp** [scalalandio/chimney#200]
Motivated by the [issues I've encountered earlier](https://github.com/scalalandio/chimney/issues/176#issuecomment-719835155), and living with `.disableDefaultValues` as a consequence, I think the following rule would be a good balance of convenience and safety: 

> Only apply default values IFF the types still match. 

To recap, my situation was that I had default values defined on a class that I was mapping to, but the type of the source mapping had changed. I was expecting a compiler error, because the types no longer matched, but because the destination had default values, those were silently applied. Very hard to find. Default values are a nice convenience, but not if they end up hiding type incompatibilities.

**Comment by nilskp** [comment]
@krzemin 
Any thoughts on this?

**Comment by krzemin** [comment]
Since 0.7.0 default values processing requires explicit enabling, so it should minimize the consequences in many codebases.

But your point is valid and we may still want to be a bit more restrictive when deciding to apply the default value.

```
case class Foo1(a: Int)
case class Foo2(a: String)
case class Target(a: Int = 1)

implicit val cfg = TransformerConfiguration.default.enableDefaultValues

Foo1(10).transformInto[Target] // should be Target(10)

Foo2("test").transformInto[Target] // should fail the compilation, but succeeds with Target(1)
```

**Comment by MateuszKubuszok** [comment]
Resolved on branch `scala-3` (#325).
